Assign array directly if type is unknown

Spea opened the pull request:
With this change it is now possible to directly assign an array to its json/model when the type is unknown. This is useful when the structure of the array is dynamic or unknown. 
In order to not break existing logic, a new configuration option `assignUnknownArrays` was added to the generator configuration.

Spea committed:
Assign array directly if type is unknown

With this change it is now possible to directly assign an array to its json/model when the type is unknown. This is useful when the structure of the array is dynamic or unknown.
In order to not break existing logic, a new configuration option `assignUnknownArrays` was added to the generator configuration.

dbu commented on src/Configuration/GeneratorConfiguration.php:
```diff
@@ -36,16 +36,23 @@ class GeneratorConfiguration implements \IteratorAggregate
+     */
+    private $assignUnknownArrays;
+
+    public function __construct(array $defaultGroupCombinations, array $defaultVersions, bool $assignUnknownArrays = false)
```
i would prefer to introduce a variable `array $options = []` here. otherwise the constructor will become cumbersome if we add more flags. currently the only supported flag will be `assign_unknown_arrays`.

we can require symfony/options-resolver (it is indirectly required anyways already) and use that to resolve the options.

Spea replied:
You are right, an options array is probably better, implemented via fe9e86955755d17308d58712b4f9a011ce97a563

dbu commented on src/Configuration/GeneratorConfiguration.php:
```diff
@@ -36,16 +36,23 @@ class GeneratorConfiguration implements \IteratorAggregate
 
-    public function __construct(array $defaultGroupCombinations, array $defaultVersions)
+    /**
+     * @var bool
```
can you please add some phpdoc here explaining what the flag does?

dbu commented on src/SerializerGenerator.php:
```diff
@@ -240,6 +240,11 @@ private function generateCodeForArray(
                 break;
 
+            case $subType instanceof PropertyTypeUnknown;
+                if ($this->configuration->shouldAssignUnknownArrays()) {
```
i think this condition needs to be part of the `case` criteria, so `case $subType instanceof PropertyTypeUnknown && $this->configuration->shouldAssignUnknownArrays():` - otherwise when the flag is false, we do not get the previous error handling.

Spea replied:
Fixed via 90786c09325b9f0daeb4fa29d9f79e0f31535a45

dbu commented on src/SerializerGenerator.php:
```diff
@@ -240,6 +240,11 @@ private function generateCodeForArray(
                 $innerCode = $this->generateCodeForClass($subType->getClassMetadata(), $apiVersion, $serializerGroups, $arrayPath.'['.$index.']', $modelPath.'['.$index.']', $stack);
                 break;
 
+            case $subType instanceof PropertyTypeUnknown;
```
```suggestion
            case $subType instanceof PropertyTypeUnknown:
```

dbu replied:
surprised that the CI does not fail with this.

dbu reviewed (changes requested):
thanks for this pull request. do i see it correctly that the field will just copy the content of the array and hope that json serializing will work?

could we also do the inverse, and assign the raw decoded json to the property in the DeserializerGenerator?

i am not sure if we need to worry about BC with this one, as previously an unknown array produced an error, so no application will rely on the behaviour currently.
given that this has the potential to mess up at runtime, i think its still right to have to enable it manually.

can you please add documentation for the flag in the [configuration section](https://github.com/liip/serializer#configuration-format) , and a note in the changelog?

Spea committed:
Move check to `case` statement

Spea committed:
Use an `options` array for configuring flags

Spea committed:
Allow passing a `GeneratorConfiguration` to the `DeserializerGenerator`

By passing the configuration, we can now access the options from this class (e.g. whether we should assign unknown arrays)

Spea committed:
Assign unknown arrays when deserializing

Spea committed:
Add new "assign unknown arrays" feature to the changelog

dbu commented on src/Configuration/GeneratorConfiguration.php:
```diff
@@ -107,9 +120,26 @@ public function getDefaultGroupCombinations(ClassToGenerate $classToGenerate): a
         }, $this->defaultGroupCombinations);
     }
 
+    public function shouldAssignUnknownArrays(): bool
```
can we phpdoc this? something like this:

```suggestion
    /**
     * If this is false, arrays with sub type PropertyTypeUnknown are treated as error.
     * If this is true, deserialize assigns the raw array and serialize just takes the raw content of the field.
     */
    public function shouldAssignUnknownArrays(): bool
```

Spea replied:
Added via 0075dd75636f3a79ed0148c5105754bd025946c3

dbu reviewed (commented):
cool, this is shaping up nicely. we are at the "naming things" bit, i think with these comments we should get ready to merge this.

i notice that we have been inconsistent with the configuration class and the DeserializerGenerator - if we had required the GeneratorConfiguration from the beginning, this would have been more elegant. but thats something for a major version change as it absolutely must be done as BC break...

Spea committed:
Change `assign_unknown_arrays` to `allow_generic_arrays`

Spea committed:
Adjust readme and php docblock

Spea committed:
Inline options parameter

Spea commented on README.md:
```diff
@@ -103,6 +106,11 @@ To generate a file without version, specify version `''` in the list of versions
+### Arrays with an unknown type
+
+If you want to directly assign arrays with an unknown type, you can do that by 
+setting the `assign_unknown_arrays` value to `true` via the `options` argument.
```
Adjusted via 0075dd75636f3a79ed0148c5105754bd025946c3

Spea commented on README.md:
```diff
@@ -103,6 +106,11 @@ To generate a file without version, specify version `''` in the list of versions
 
+### Arrays with an unknown type
+
+If you want to directly assign arrays with an unknown type, you can do that by 
```
Adjusted via 0075dd75636f3a79ed0148c5105754bd025946c3

Spea commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+# 2.2.0 (unreleased)
+
+* Add new parameter `$options` to the `GenerateConfiguration` class
+* Support (de)serializing unknown arrays by setting the `assign_unknown_arrays`
```
Yeah that makes indeed more sense now. Adjusted this via 331604d2dbe346739ec57f5babf6d7e76b07963e

dbu commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+# 2.2.0 (unreleased)
+
+* Add new parameter `$options` to the `GenerateConfiguration` class
+* Support (de)serializing unknown arrays by setting the `allow_generic_arrays`
```
```suggestion
* Support (de)serializing arrays with undefined content by setting the `allow_generic_arrays`
```

Spea replied:
Adjusted via 02366203b17264dfb3c3468e90fea3fb87feb1a7

dbu reviewed (approved):
awesome, thanks a lot for this improvement!

Spea committed:
Adjust wording in changelog

dbu commented:
thanks for the contribution! should i tag 2.2 right away or do you have more features you plan on doing soon? have you used this feature in an application already?

Spea commented:
> thanks for the contribution! should i tag 2.2 right away or do you have more features you plan on doing soon? have you used this feature in an application already?

As of right now I have no more features planned. I already used this feature in one of our applications. However, it might make sense to wait a few days/weeks before doing a new release. I can use the master branch in the meantime.

dbu commented:
i tagged https://github.com/liip/serializer/releases/tag/2.2.0

and i renamed master to 2.x to be more explicit with versions and if in the future you need latest, you would not risk random BC breaks like when requiring master.
